Add NFA-based regular expression utility in CPC

ofecisrael opened the pull request:
This pull request introduces new CPC programs for evaluating constant string membership in regular expressions using NFA-based algorithm, providing a more efficient alternative to the existing evaluation method.

ajreynol commented on proofs/eo/cpc/programs/Nfa.eo:
```diff
@@ -0,0 +1,155 @@
+; Declares a new sort which will be used to represent the states and transitions
+; of the automaton.
+(declare-type Nfa ())
```
Can you add `@` as a prefix of all types/consts introduced? So that `@nfa.` is the prefix for each of those below and `@Nfa` is the type.

ajreynol commented on include/cvc5/cvc5_proof_rule.h:
```diff
@@ -3773,6 +3773,20 @@ enum ENUM(ProofRewriteRule)
+   *
+   * \endverbatim
+   */
+  EVALUE(STR_IN_RE_EVAL_NFA),
```
I think I'd prefer not (temporarily) changing the API with this change. 

What I'd recommend is: 
- We merge this PR with the new Eunoia implementation, but have it unable to be activated for now.
- I will test on my end the addition (to confirm the improvement) and then replace the current implementation with the new one.  Then, I will replace the Eunoia signature for str_in_re_eval with your implementation and send the PR that activates it.

Sound good?

ajreynol commented on proofs/eo/cpc/rules/Nfa.eo:
```diff
@@ -0,0 +1,14 @@
+(include "../programs/Nfa.eo")
```
As mentioned, let's omit `rules/Nfa.eo`.  I will replace the call to your method on a separate PR.

ajreynol commented on src/options/proof_options.toml:
```diff
@@ -269,3 +269,11 @@ name   = "Proof"
   default    = "2"
   help       = "the lookahead depth for merging proofs at previsit"
+
+[[option]]
```
Since it's best not to modify the API, I think we have to omit this option.

For testing purposes, I would maintain this option on a dev branch however.

ajreynol commented on test/regress/cli/regress0/proofs/dsl-no-eval.smt2:
```diff
@@ -1,3 +1,5 @@
+; COMMAND-LINE:
+; COMMAND-LINE: --proof-cpc-str-in-re-nfa
```
Ditto, let's not modify the regressions.

ajreynol commented on proofs/eo/cpc/programs/Nfa.eo:
```diff
@@ -0,0 +1,155 @@
+; the just created NFA for "b" as its tail. This elegantly links the automaton parts together.
+; return: The Nfa that was built from the regex.
+(program $build_nfa ((tail Nfa) (s String) (r1 RegLan) (rr RegLan :list) (c0 String) (c1 String))
+  (RegLan Nfa) Nfa
```
Can you add the new `:signature` keyword before the arg types / return type of each of the programs? (We introduced this recently to improve readiability).
```suggestion
  :signature (RegLan Nfa) Nfa
```

ajreynol reviewed (changes requested):
Thanks a lot for the PR, looking forward to having this in main :)

I have some requests for changes, I think its best to not temporarily change the API.  I will send a PR shortly after this merges to activate your method.

Also, can you quickly confirm that your method is fast for `regress0/strings/re-mem-eval-large.smt2`?  This regression is currently disabled for CPC since the current method does not scale. I will reenable it when I send the followup PR.

ofecisrael commented:
> Thanks a lot for the PR, looking forward to having this in main :)
> 
> I have some requests for changes, I think its best to not temporarily change the API. I will send a PR shortly after this merges to activate your method.
> 
> Also, can you quickly confirm that your method is fast for `regress0/strings/re-mem-eval-large.smt2`? This regression is currently disabled for CPC since the current method does not scale. I will reenable it when I send the followup PR.

Yes, I can confirm that it runs in ~0.4 seconds on my machine.

ofecisrael commented:
> Can you update the PR description? Also, can you ensure that `Nfa.eo` has the appropriate includes so that ethos can parse it standalone?

I've updated the PR description and fixed the includes in Nfa.eo. I also updated the includes in theories/Quantifiers.eo to ensure it has all the necessary dependencies as well.

ofecisrael commented:
> @ofecisrael I've opened https://github.com/cvc5/cvc5/pull/12025 which has the integration of your side condition into the core utility.  (Note that I integrated your code into the core utility and fallback to the internal method there, not the other way around).  If this looks good to you, then let's go ahead and merge this PR.

Sounds great! The other PR looks good to me as well.

ajreynol reviewed (approved):
LGTM, excited to have this on main!
